Pull services in parallel

edsrzf opened the pull request:
Updates #1652

This PR adds a new `--parallel` flag on `docker-compose pull` that pulls images in parallel, which should hopefully speed up pulls for projects with lots of services. By default, with no flag, you get the old, non-parallel behavior.

`parallel_execute` now accepts an optional `limit` parameter that specifies the number of threads that actually run at one time. There are arguably more elegant approaches than mine, but I opted to keep the diff relatively small. I'm happy to get some feedback about this. I've somewhat arbitrarily chosen 5 as the limit for the number of parallel pulls. I'm also happy to break this part out into a separate PR since it's mostly independent from the rest of this PR and could be useful in other contexts.

The normal `docker-compose pull` output is silenced when pulling in parallel. This means we no longer get progress bars with `--parallel`, but the output is easier to understand.

I've tested this manually and it seems to work. I'm not sure how I'd go about
writing an automated test for it, but open to suggestions.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "parallel-pull" git@github.com:edsrzf/compose.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "parallel-pull" git@github.com:edsrzf/compose.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353904296
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

shin- commented:
Thanks! 

I'm honestly not sure how well this would scale, how the engine handles 10, maybe 20 simultaneous pulls - would this potentially cause timeouts or significant slowdowns on certain networks - how well does our client handle maintaining several connections open at the same time - not to mention it probably messes up the output in Compose.

In order for me to accept this, it would at least need to be an option behind a flag, e.g `docker-compose pull --parallel`, and disabled by default.

edsrzf commented:
Thanks for the feedback! I want to do this right. I'm fine with keeping it behind a `--parallel` flag to start with, but would also like to try and fix the problems so that this could, perhaps someday, be the default.

Admittedly, the output is a bit weird. If we could limit each pull to a single status line and maybe get rid of the progress bars, or at least confine them to one line, that might be more manageable.

I've been running with this on a project that has 27 different services and been having no real problems pulling all images in parallel, but I can imagine that it could cause problems with some setups. Maybe we could limit the maximum number of pulls running at a time?

I'll look at `parallel.py` and see what I can do for both of these things.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "parallel-pull" git@github.com:edsrzf/compose.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354376568
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

edsrzf commented:
I've updated. Here's what's new:

There's a `--parallel` flag on `docker-compose pull`. By default, you get the old, non-parallel behavior.

`parallel_execute` now accepts an optional `limit` parameter that specifies the number of threads that actually run at one time. There are arguably more elegant approaches than mine, but I opted to keep the diff relatively small. I'm happy to get some feedback about this. I've somewhat arbitrarily chosen 5 as the limit for the number of parallel pulls. I'm also happy to break this part out into a separate PR since it's mostly independent from the rest of this PR and could be useful in other contexts.

The normal `docker-compose pull` output is silenced when pulling in parallel. This means we no longer get progress bars with `--parallel`, but the output is easier to understand.

shin- commented on compose/project.py:
```diff
@@ -454,9 +454,22 @@ def _get_convergence_plans(self, services, strategy):
-    def pull(self, service_names=None, ignore_pull_failures=False):
-        for service in self.get_services(service_names, include_deps=False):
-            service.pull(ignore_pull_failures)
+    def pull(self, service_names=None, ignore_pull_failures=False, in_parallel=False):
```
Nit: can we rename `in_parallel` to `parallel_pull` instead?

shin- commented on compose/service.py:
```diff
@@ -885,17 +885,19 @@ def has_host_port(binding):
         return any(has_host_port(binding) for binding in self.options.get('ports', []))
 
-    def pull(self, ignore_pull_failures=False):
+    def pull(self, ignore_pull_failures=False, silent=False):
```
Maybe instead of `silent`, have a `no_progress` argument that still logs the `Pulling ...` lines, but omits the progress_stream part.

edsrzf replied:
There are a couple problems with doing this:

First, `parallel_execute` outputs lines very similar to `Service.pull`. It's

`Pulling service (image:tag) ...`

vs.

`Pulling service ...`

So we get the same thing output twice for each service, basically.

Second, `parallel_execute` outputs some terminal commands to go back and update older lines, and if anything else outputs something it messes up the terminal commands and the wrong lines are updated.

shin- replied:
Nevermind then - as long as we have some level of feedback for the user during the operation, that should be sufficient.

shin- commented on compose/service.py:
```diff
@@ -885,17 +885,19 @@ def has_host_port(binding):
-            return progress_stream.get_digest_from_pull(
-                stream_output(output, sys.stdout))
+            if not silent:
+                return progress_stream.get_digest_from_pull(
```
I think we still need the result of `get_digest_from_pull` here, but we need to modify the  `stream_output` call to not write to `stdout`. Maybe something like
```python
with open(os.devnull, 'w') as devnull:
  return progress_stream.get_digest_from_pull(
    stream_output(output, devnull)
  )
```

edsrzf replied:
I've got no problem doing this, but as far as I can tell no one uses the value returned from this method.

shin- replied:
The cost of doing this seems small enough - it might be used by people using Compose as a library, or we might want to use it in the future - so I'd say it's worth it.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "parallel-pull" git@github.com:edsrzf/compose.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353574984
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

edsrzf commented:
Updated! Issues from code review should be addressed. I also changed from using a `Queue` to a `Semaphore` to limit parallelism since it just made more sense. And finally, I added a test for `parallel_execute` when there's a limit.

Please take another look.
